src/boot/bootctl-install.c

```diff
@@ -423,12 +460,192 @@ static int copy_one_file(const char *esp_path, const char *name, bool force) {
+            !startswith_no_case(de->d_name, "boot"))
+                return RECURSE_DIR_CONTINUE;
+
+        _cleanup_free_ char *csv = NULL;
+        size_t csv_size = 0;
+        r = read_full_file(source_path, &csv, &csv_size);
+        if (r < 0) {
+                log_error_errno(r, "Failed to read %s: %m", path);
+                return RECURSE_DIR_CONTINUE;
+        }
+
+        _cleanup_free_ char *file = NULL, *desc = NULL;
+        r = parse_boot_csv_mem((char16_t*)csv, csv_size, &file, &desc);
+        if (r < 0)
+                return RECURSE_DIR_CONTINUE;
+
+        _cleanup_free_ char *dir = NULL;
+        if (path_extract_directory(path, &dir) < 0)
+                return RECURSE_DIR_CONTINUE;
+
+        _cleanup_free_ char *p = path_join(dir, file);
+        if (!p)
+                return log_oom();
+
+        log_debug("Found default loader %s in %s", p, source_path);
+        *(args->loader_path) = TAKE_PTR(p);
+        *(args->loader_description) = TAKE_PTR(desc);
+
+        return RECURSE_DIR_CONTINUE;
+}
+
+static int install_binaries_extra(
+                const char *esp_path,
+                const char *arch,
+                bool force,
+                char **ret_loader_path,
+                char **ret_loader_description) {
+
+        char *root = IN_SET(arg_install_source, ARG_INSTALL_SOURCE_AUTO, ARG_INSTALL_SOURCE_IMAGE) ? arg_root : NULL;
+        _cleanup_closedir_ DIR *d = NULL;
+        _cleanup_free_ char *path = NULL, *loader_path = NULL, *loader_description = NULL;
+        int r;
+
+        r = chase_and_opendir(BOOTLIBDIR "/extra", root, CHASE_PREFIX_ROOT|CHASE_PROHIBIT_SYMLINKS, &path, &d);
+        /* If we had a root directory to try, we didn't find it and we are in auto mode, retry on the host */
+        if (r == -ENOENT && root && arg_install_source == ARG_INSTALL_SOURCE_AUTO)
+                r = chase_and_opendir(BOOTLIBDIR "/extra", NULL, CHASE_PREFIX_ROOT|CHASE_PROHIBIT_SYMLINKS, &path, &d);
+        if (r == -ENOENT) {
+                if (ret_loader_path)
+                        *ret_loader_path = NULL;
+                if (ret_loader_description)
+                        *ret_loader_description = NULL;
+                return 0; /* ignore if vendor dir does not exist */
+        }
+        if (r < 0)
+                return log_error_errno(r, "Failed to open " BOOTLIBDIR "/extra: %m");
+
+        r = recurse_dir(dirfd(d), NULL, 0, 3, RECURSE_DIR_IGNORE_DOT, copy_loader,
+                        &(struct CopyLoaderArgs) {
+                                .source_path = BOOTLIBDIR "/extra",
+                                .esp_path = esp_path,
+                                .force = force,
+                                .loader_path = &loader_path,
+                                .loader_description = &loader_description,
+                        });
+        if (r < 0)
+                return log_error_errno(r, "Failed to recurse %s: %m", path);
```

Comment:
btw, what's the rationale for using your own recurse_dir() logic here if you are not going to use the version check stuff anyway? because then you could just use copy_tree_at() and maybe just add a flag there that suppresses copies of files whose target already matches the source's contents.

That sounds much preferable and simpler.